card/lib/card/query.rb

```diff
@@ -129,38 +129,84 @@ def default_comment
     # run the current query
     # @return array of card objects by default
     def run
-      retrn ||= statement[:return].present? ? statement[:return].to_s : "card"
-      if retrn == "card"
-        get_results("name").map do |name|
-          Card.fetch name, new: {}
-        end
-      else
-        get_results retrn
+      retrn = statement[:return].present? ? statement[:return].to_s : "card"
+      return_method = :"return_#{simple_result?(retrn) ? :simple : :list}"
+      send return_method, run_sql, retrn
+    end
+
+    # @return [(not an Array)]
+    def return_simple sql_result, retrn
+      send result_method(retrn), sql_result, retrn
+    end
+
+    # @return [Array]
+    def return_list sql_results, retrn
+      sql_results.map do |record|
+        return_simple record, retrn
+      end
+    end
+
+    def result_method retrn
+      case
+      when respond_to?(:"#{retrn}_result") then :"#{retrn}_result"
+      when (retrn =~ /id$/)                then :id_result
+      when (retrn =~ /_\w+/)               then :name_result
+      else                                      :default_result
       end
     end
 
-    # @return Integer for :count, otherwise Array of Strings or Integers
-    def get_results retrn
-      rows = run_sql
-      if retrn =~ /_\w+/
-        name_processor = contextual_name_processor retrn
-        rows.map do |row|
-          name_processor.call row["name"]
-        end
-      elsif retrn == "name" && (statement[:prepend] || statement[:append])
-        rows.map do |row|
-          [statement[:prepend], row["name"], statement[:append]].compact * "+"
-        end
+    def count_result results, _field
+      results.first["count"].to_i
+    end
+
+    def default_result record, field
+      record[field]
+    end
+
+    def id_result record, field
+      record[field].to_i
+    end
+
+    def raw_result record, _field
+      record
+    end
+
+    def simple_result? retrn
+      retrn == "count"
+    end
+
+    def name_result record, pattern
+      process_name record["name"], pattern
+    end
+
+    def process_name name, pattern
+      name = pattern.to_name.absolute(name) if pattern =~ /_\w+/
```

Comment:
the biggest conflict to resolve since I last worked on this was with the name contextualization, especially the "contextual name processor".  It basically seems to have done the same thing as `pattern.to_name.absolute(name)`, however (as you can see in the spec), it would return a card with the name `""`, whereas the behavior of `#absolute` (for better or for worse) is just to return the context if the name doesn't match the context. 

For now I changed the spec, but I'm not sure what we really want.  Are we ok with queries returning cards with no name?

In either case, I think we probably want that processed in Card::Name as it now is.  But we might want to reconsider (or add more options to) the absolutization behavior